### PR TITLE
add u256 multiplication

### PR DESCRIPTION
**Pull request by peterjah**
I took the logic from this golang lib:
https://github.com/holiman/uint256
it is the library used in go-ethereum to handle uint256 operation

### CODE REVIEW

File: assembly/globals.ts [hunk 1]
```diff
@@ -75,14 +78,73 @@ export function __umulq64(a: u64, b: u64): u64 {
...
+@global
+export function __mul256(x0: u64, x1: u64, x2: u64, x3: u64, y0: u64, y1: u64, y2: u64, y3: u64): u256 {
+  var lo1 = __umulq64(x0, y0);
+  var res1 = __umul64Hop(__res128_hi, x1, y0);
+  var res2 = __umul64Hop(__res128_hi, x2, y0);
+  var res3 = x3 * y0 + __res128_hi;
+
+  var lo2 = __umul64Hop(res1, x0, y1);
+  res2 = __umul64Step(res2, x1, y1, __res128_hi);
+  res3 = res3 + x2*y1 + __res128_hi;
```
Comment (MaxGraey) [inline assembly/globals.ts]:
```suggestion
  res3 += x2 * y1 + __res128_hi;
```

File: assembly/globals.ts [hunk 2]
```diff
@@ -75,14 +78,73 @@ export function __umulq64(a: u64, b: u64): u64 {
...
+
+  var hi1 = __umul64Hop(res2, x0, y2);
+  res3 = res3 + x1*y2 + __res128_hi
```
Comment (MaxGraey) [inline assembly/globals.ts]:
```suggestion
  res3 += x1 * y2 + __res128_hi
```

File: assembly/globals.ts [hunk 3]
```diff
@@ -75,14 +78,73 @@ export function __umulq64(a: u64, b: u64): u64 {
...
-      uv = u * b + (uv & 0xFFFFFFFF);
+  uv = u * b + (uv & 0xFFFFFFFF);
 
   __res128_hi = a * b + w1 + (uv >> 32);
   return (uv << 32) | w0;
 }
 
+// __umul64Hop computes (hi * 2^64 + lo) = z + (x * y)
+// @ts-ignore: decorator
+@global
```
Comment (MaxGraey) [inline assembly/globals.ts]:
```suggestion
@inline
```

File: assembly/globals.ts [hunk 4]
```diff
@@ -75,14 +78,73 @@ export function __umulq64(a: u64, b: u64): u64 {
...
+  var lo = __umulq64(x, y);
+  lo = __uadd64(lo, z);
+  var hi = __res128_hi +__carry;
+  __res128_hi = hi;
+  return lo
+}
+
+// __umul64Step computes (hi * 2^64 + lo) = z + (x * y) + carry.
+// @ts-ignore: decorator
+@global
```
Comment (MaxGraey) [inline assembly/globals.ts]:
```suggestion
@inline
```

File: assembly/globals.ts [hunk 5]
```diff
@@ -75,14 +78,73 @@ export function __umulq64(a: u64, b: u64): u64 {
...
+  hi += __carry;
+  __res128_hi = hi;
+  return lo
+}
+
+// __uadd64 returns the sum with carry of x, y and carry: sum = x + y + carry.
+// The carry input must be 0 or 1; otherwise the behavior is undefined.
+// The carryOut output is guaranteed to be 0 or 1.
+// @ts-ignore: decorator
+@global
```
Comment (MaxGraey) [inline assembly/globals.ts]:
```suggestion
@inline
```

File: assembly/__tests__/u256.spec.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import { u128 } from '../integer';
```
Comment (MaxGraey) [inline assembly/__tests__/u256.spec.ts]:
```suggestion
import { u128 } from '../integer/u128';
```

File: assembly/__tests__/u256.spec.ts [hunk 7]
```diff
@@ -348,4 +349,74 @@ describe("Basic Operations", () => {
...
+      const a = u256.from(randa);
+      const randb = <u64>Math.trunc(Math.random() * 0xFFFF)
+      const b = u256.from(randb);
+      const r = new u256(randa * randb);
+
+      expect(a * b).toStrictEqual(r);
+      expect(b * a).toStrictEqual(r);
+      i++;
+    }
+  });
```
Comment (MaxGraey) [inline assembly/__tests__/u256.spec.ts]: This randomly tests only lowest `u64` part with range `0 - 0xFFFF`. Which doesn't make sense to me

Comment (MaxGraey) [reply assembly/__tests__/u256.spec.ts]: It would be nice to have some manual (not randomized) tests for u256 that cover carry bits in all nibs (u64 parts). And also tests with overflow of final result

**Comment by MaxGraey** [comment]
Could you also mention original `https://github.com/holiman/uint256` in comments for __mul256?
Otherwise LGTM.

**Comment by MaxGraey** [comment]
Thanks!
